ansible_base/oauth2_provider/views/permissions.py

```diff
@@ -0,0 +1,35 @@
+from rest_framework import permissions
+
+
+class OAuth2TokenPermission(permissions.BasePermission):
+    # An app token is a token that has an application attached to it
+    # A personal access token (PAT) is a token with no application attached to it
+    # With that in mind:
+    # - An app token can be read, changed, or deleted if:
+    #   - I am the superuser
+    #   - I am the admin of the organization of the application of the token
+    #   - I am the user of the token
+    # - An app token can be created if:
+    #   - I have read access to the application (currently this means: I am the superuser)
+    # - A PAT can be read, changed, or deleted if:
+    #   - I am the superuser
+    #   - I am the user of the token
+    # - A PAT can be created if:
+    #   - I am a user
+
+    def has_permission(self, request, view):
+        # Handle PAT and app token creation separately
+        if request.method == 'POST':
+            if request.data.get('application'):
+                # TODO: Change this once dab#424 is fixed
+                return request.user.is_superuser
+        return request.user.is_authenticated
+
+    def has_object_permission(self, request, view, obj):
+        if request.method in ['GET', 'PATCH', 'PUT', 'DELETE']:
+            if request.user.is_superuser:
+                return True
+            if obj.application and obj.application.organization.admins.filter(id=request.user.id).exists():
```

Comment:
Isn't this in-congruent with the creation permission (`has_permission`)?

If we require permission to the related application of a token (for the type of tokens that have an application) then creation would require organization admin, and change permission would require organization admin.

If we can't support that easily for creation, I think it might make more sense to leave both create/edit at superuser level (for type of tokens that have an application) and leave the full implementation as a TODO. Otherwise a half-way solution can cause even more confusion.

Reply:
This is knowingly not complete, as per test `xfail`, so resolving this for now. I will not block merge on org admins not being able to create tokens, and we can leave it this way.